### ISSUE TITLE
"SyntaxError: Use of const in strict mode" on Node 0.12

### ISSUE THREAD

**Issue by Turbo87** [rollup/rollup#909]
After trying to upgrade to rollup 0.35.0 in https://github.com/ember-cli/ember-cli-lodash-subset/pull/13 we noticed that it doesn't seem to work on Node 0.12 anymore as you can see on the Travis builds on that PR.

specifically we're seeing this:

```
/home/travis/build/ember-cli/ember-cli-lodash-subset/node_modules/rollup/dist/rollup.js:18
const absolutePath = /^(?:\/|(?:[A-Za-z]:)?[\\|\/])/;
^^^^^
SyntaxError: Use of const in strict mode.
    at exports.runInThisContext (vm.js:73:16)
    at Module._compile (module.js:443:25)
    at Object.Module._extensions..js (module.js:478:10)
    at Module.load (module.js:355:32)
    at Function.Module._load (module.js:310:12)
    at Module.require (module.js:365:17)
    at require (module.js:384:17)
    at Object.<anonymous> (/home/travis/build/ember-cli/ember-cli-lodash-subset/node_modules/rollup/bin/rollup:5:14)
    at Module._compile (module.js:460:26)
    at Object.Module._extensions..js (module.js:478:10)
```

**Comment by Rich-Harris** [comment]
Gah. I didn't think supporting Node 0.12 was any more important than supporting IE8 at this time. People might still be running their *apps* on 0.12 (though frankly they shouldn't be) but is anyone really *building* stuff or running CLI tools on 0.12?

How important is it to reinstate 0.12 support? Rollup 0.35 makes liberal use of things that I don't think are fully supported on 0.12, like `for-of` loops. Would be a nuisance to have to introduce a dependency on Babel just so we can support a tiny minority of holdouts.

**Comment by snuggs** [comment]
This is true @Rich-Harris. Any way we can get some type of metric on Node 0.12 usage? Not that this would change the plan of action. More out of my own curiosity. @Turbo87

**Comment by Turbo87** [comment]
the reason why I'm bringing this up is detailed in http://emberjs.com/blog/2016/09/07/ember-node-lts-support.html 

I'm personally using node 6 but unfortunately we still seem to have some users that are stuck on 0.x...

**Comment by snuggs** [comment]
@Rich-Harris couldn't we create a `rollup-plugin-strict-mode` plugin that handles conversion to vars? I may be oversimplifying the situation but that's clearly what we want. Your thoughts @Turbo87 /cc @TrySound @Victorystick

**Comment by Rich-Harris** [comment]
@snuggs the problem isn't the code Rollup generates, it's that Rollup straight up doesn't run in 0.12. 

> Any way we can get some type of metric on Node 0.12 usage?

I googled but didn't find anything. At any rate it would be unreliable because it would presumably include 0.12 deployments, which aren't the issue here.

As I see it there are three options in front of us:

1. Ember libraries are forced to stay on 0.34 until Jan 1 2017, when 0.12 support is officially dropped
2. We create a separate `rollup-legacy` package that basically just Babelifies Rollup
3. We see what it would take to continue supporting 0.12, at least until Jan 1

None are ideal. 3 is probably the least worst.

Ember's stance is commendable but I maintain that the users we're talking about are imaginary!

**Comment by TrySound** [comment]
Four months is nothing. After that still somebody will come and ask to support old stuff. My vote is first variant.

**Comment by Turbo87** [comment]
@Rich-Harris I think the issue is essentially that v0.34 worked with Node 0.12 while v0.35 doesn't but I couldn't find any information on Node version compat in the readme or the changelog about it. If you communicate clearly that rollup is only compatible with Node 4.x and above we're probably fine.

**Comment by nathanhammond** [comment]
Good morning everybody! I figure I should chime in [as the author of the aforementioned post](http://emberjs.com/blog/2016/09/07/ember-node-lts-support.html) and another of the Ember CLI maintainers.

This decision is something that we elected to make it easier to support our enterprise customers who desire predictability and guarantees. @Rich-Harris I know for a fact that the users we're talking about aren't imaginary; three of the major participants in our ecosystem are all still running 0.X versions of Node.js: Ghost, LinkedIn, and Yahoo.

Our approach inside of the Ember community is presently the most clear statement of support for the Node.js LTS process. That our position is contentious speaks directly to the need we have as a community to validate the legitimacy of the Node.js LTS plans. At the very least we'd like for Rollup to take this into account for planning [support for 4.X](https://github.com/nodejs/LTS#lts-schedule).

All three options proposed are reasonable, but Rollup is a bit different for us than [ESLint dropping support](https://github.com/eslint/eslint/issues/6356#issuecomment-229372009): it more-often shows up as a transitive dependency given our addon story. As such, Option 1 may be difficult to support (though maybe possible, considering the time lag before it came to our attention) and Option 2 will result in our attempt to move much of an ecosystem to a different package of Rollup which likely has worse long-term consequences.

Unsurprisingly I'm in favor of Option 3 though the [number of changes seem daunting](https://github.com/rollup/rollup/compare/v0.34.13...v0.35.7).

**Comment by Rich-Harris** [comment]
@Turbo87 

> I couldn't find any information on Node version compat in the readme or the changelog about it

Yeah, that's my bad – sorry! I was in the midst of a big rewrite, saw it in the `.travis.yml`, briefly thought 'what is that doing there?' as I deleted it then promptly forgot about it.

@nathanhammond Thanks –

> three of the major participants in our ecosystem are all still running 0.X versions of Node.js

I've no doubt that people are still running apps on 0.X – am just surprised if people are *building* those apps on 0.X. I hope they're able to upgrade soon for their sake! I have shivers when I think back to that time.

Anyway, good news – turns out it might not be such a big deal after all. The changes required for Rollup to support 0.12 are extremely minor (basically just reconfiguring Bublé to transpile to 0.12 instead of 4), so I've opened #909.

**Comment by nathanhammond** [comment]
@Rich-Harris LinkedIn is my employer and I can confirm that we're using 0.12 for builds. Not particularly ideal and planning to move by the end of the year... but we've prioritized it with all of our other work. It's why having set timelines is important for us.

Thank you for investigating; I'm glad that #927 appears so minimal!

**Comment by eventualbuddha** [comment]
At Square we just moved from 0.12 (and 0.10 in some cases!) to 5.10. Not a terrible move but definitely some work.

**Comment by Rich-Harris** [comment]
Have released 0.35.9 with 0.12 support, and Travis/Appveyor config to prevent regressions. Thanks all

**Comment by snuggs** [comment]
Another one bites the dust! @stefanpenner
